### ISSUE TITLE
[x64/MacOs] Memory corruption with optimisations enabled.

### ISSUE THREAD

**Issue by BastianBlokland** [dotnet/coreclr#24657]
Hi,

I'm facing an issue where a specific scenario causes the runtime to write outside of the struct boundary when run with optimisations enabled.

Repro:
```c#
public class Program
{
    struct Wrapper<T>
    {
        public T Val { get; set; }

        public static implicit operator Wrapper<T>(T val) => new Wrapper<T> { Val = val };
    }

    struct TestStruct
    {
        public Wrapper<int> Field1;
        public Wrapper<short> Field2;
        public Wrapper<short> Field3;

        public TestStruct(int field1)
        {
            this.Field1 = field1;
            this.Field2 = 0;
            this.Field3 = 0;
        }
    }

    public static void Main(string[] args)
    {
        var array = new [] { new TestStruct(123), new TestStruct(456) };
        if (array[1].Field1.Val != 456)
            throw new System.Exception("Does not happen");

        array[0].Field3 = 100;

        if (array[1].Field1.Val != 456)
            throw new System.Exception("Corrupt: " + array[1].Field1.Val);

        System.Console.WriteLine("Success");
    }
}
```
Sorry for the slightly convoluted repro but this seems to be the smallest i can have it before the issue disappears.
 
Output when running that in a fresh `netcoreapp2.2` console project on mac-os:
`dotnet run -c debug`: `Success`
`dotnet run -c release`: `Unhandled Exception: System.Exception: Corrupt: 0`

This issues doesn't seem to exist on `Windows` but i've seen similar failures on `Linux`.

Greetings,
Bastian

**Comment by CarolEidt** [comment]
I am able to repro this on Linux with 2.2. It doesn't repro with latest. I'm working on tracking down when/how it was fixed.

**Comment by CarolEidt** [comment]
This seems to be a duplicate of https://github.com/dotnet/coreclr/issues/18522. I'm working on verifying that.

**Comment by CarolEidt** [comment]
This was indeed a duplicate of #18522. The fix will be in 3.0.

**Comment by BastianBlokland** [comment]
Thank you for investigating, looking forward to upgrading to 3.0 👍
